[ENH]: CIP-5 - Large Batch Handling Improvements Proposal

tazarov opened the pull request:
- Including only CIP for review.

Refs: #1049

## Description of changes

*Summarize the changes made by this PR.*
 - Improvements & Bug fixes
	 - New proposal to handle large batches of embeddings gracefully

## Test plan
*How are these changes tested?*

- [ ] Tests pass locally with `pytest` for python, `yarn test` for js

## Documentation Changes
TBD

HammadB commented on docs/CIP_5_Large_Batch_Handling_Improvements.md:
```diff
@@ -0,0 +1,56 @@
+  and `chromadb.api.fastapi.FastAPI`
+- `chromadb.api.segment.SegmentAPI` will implement the `max_batch_size` property by fetching the value from the
+  `Producer` class.
+- `chromadb.api.fastapi.FastAPI` will implement the `max_batch_size` by fetching it from a new `/pre-flight-checks`
```
is the assumption that these are fetched once on client creation?

tazarov replied:
Haven't thought about it but actual impl caches the values

tazarov replied:
https://github.com/amikos-tech/chroma-core/blob/93f0eda92ff1061a42cb6b3cf219ce6291c6f4ef/chromadb/api/fastapi.py#L400

HammadB commented on docs/CIP_5_Large_Batch_Handling_Improvements.md:
```diff
@@ -0,0 +1,56 @@
+The following interfaces are impacted:
+
+- New Server API endpoint - `/pre-flight-checks`
+- New `max_batch_size` property on the `API` interface
```
Naming thought: Is it confusing that this only applies to add?

tazarov replied:
Upsert too, gonna add that. How about get/delete - maybe future PR?

tazarov replied:
added it to all three - add, update, upsert

tazarov commented:
@HammadB, true that we cannot guarantee ACID-like behaviour this way, but even without, things seem to go wrong on very large batches. Do you think we can introduce `tenacity` (Apache 2 lic) for some added bonus of retries? I think this can be helpful further down the road, where a retry can resolve the problem.

In my experience, in situations where transactionality cannot be guaranteed, we can implement mechanisms to fail gracefully - e.g. store the failed batch in a temp file and allow the user to resubmit later. wdyt?

imartinez commented:
I think this discussion brought by @HammadB and the last proposed solution by @tazarov requires more time and analysis (the last solution also has its implications). 

I'd suggest to reduce the scope of this change to just exposing the max_batch_size in the API so it can be merged and clients can handle batches themselves (and add it to the docs) -in the case of privateGPT it is causing errors, I guess there are more clients impacted- and discuss the error management in a different CIP. WDYT?

tazarov commented:
@imartinez, let me do that now.

tazarov commented:
@imartinez @HammadB, Refactored. The new impl no longer splits batches automatically. We do ship a utility library for that though `chromadb/utils/batch_utils.py`. The API now validates the batch size (client-side), prior to trying to persist it. Everything else is left as-is:

- pre-flight-check endpoint is still there
- max_batch_size property is still available at API level for clients to use to break up batches

Added a couple more tests to ensure correctness.

CIP doc has also been updated to reflect above changes.

imartinez commented on chromadb/utils/batch_utils.py:
```diff
@@ -0,0 +1,34 @@
+)
+
+
+def create_batches(
```
It'd be extra convenient from a client perspective to be able to pass here the API instead of the max_batch_size, and have the function extract the max_batch_size from the API object. That way the client doesn't need to call the API to get the max_batch_size to then pass it to this function. Saves one line of code ;) 

It is just a nice to have... maybe solvable by "overloading" this function.

tazarov replied:
Done and tested.

HammadB replied:
seems like this change wasn't made?

tazarov replied:
I had made it but the rebase I did from main screw it up - https://github.com/amikos-tech/chroma-core/blob/c0d42f3dc8110371fc1a49d7db17af7e75cbbc8b/chromadb/utils/batch_utils.py

Reverting now.

tazarov commented:
@HammadB,

Another unrelated failure:

```bash
chromadb/test/property/test_persist.py .F                                [100%]

=================================== FAILURES ===================================
___________________ test_persist_embeddings_state[settings0] ___________________

caplog = <_pytest.logging.LogCaptureFixture object at 0x7f496ee70550>
settings = Settings(environment='', chroma_db_impl=None, chroma_api_impl='chromadb.api.segment.SegmentAPI', chroma_telemetry_impl...=None, chroma_server_auth_token_transport_header=None, anonymized_telemetry=True, allow_reset=True, migrations='apply')

    def test_persist_embeddings_state(
        caplog: pytest.LogCaptureFixture, settings: Settings
    ) -> None:
        caplog.set_level(logging.ERROR)
        api = chromadb.Client(settings)
        run_state_machine_as_test(
>           lambda: PersistEmbeddingsStateMachine(settings=settings, api=api)
        )  # type: ignore

chromadb/test/property/test_persist.py:226: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/hypothesis/stateful.py:216: in run_state_machine_as_test
    run_state_machine(state_machine_factory)
/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/hypothesis/stateful.py:112: in run_state_machine
    @given(st.data())
chromadb/test/property/test_embeddings.py:210: in ann_accuracy
    embedding_function=self.embedding_function,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

collection = Collection(name=A01
)
record_set = {'documents': [None], 'embeddings': [[0.09765625, 0.430419921875, 0.20556640625, 0.08978271484375, -0.1527099609375, 0.2[917](https://github.com/chroma-core/chroma/actions/runs/6082759956/job/16515250514?pr=1077#step:5:918)48046875, ...]], 'ids': ['1'], 'metadatas': [{'1': '10'}]}
n_results = 1, min_recall = 0.95
embedding_function = <chromadb.test.property.strategies.hashing_embedding_function object at 0x7f4956ddcc50>
```

tazarov commented:
@HammadB, can you rerun the failed tests just to see if we can rule out intermittent failure? If it keeps failing, I'll debug it further

HammadB commented:
@tazarov forsure, I've rerun them a handful of times but happy to do so again

tazarov commented:
@HammadB, Tests in my fork are passing, so this is ready for merge after your review.

HammadB commented on chromadb/api/fastapi.py:
```diff
@@ -34,10 +36,12 @@
 from urllib.parse import urlparse, urlunparse, quote
 
 logger = logging.getLogger(__name__)
+from chromadb.utils.batch_utils import create_batches
```
unused?

HammadB commented on chromadb/api/fastapi.py:
```diff
@@ -434,6 +438,15 @@ def get_settings(self) -> Settings:
+    @property
+    @override
+    def max_batch_size(self) -> int:
+        resp = self._session.get(self._api_url + "/pre-flight-checks")
```
why are we making the request everytime? If we have it cached, we can safely ignore the get?

tazarov replied:
I thought I'd fixed it. Fixed now.
